miniapps/spde/spde_solver.cpp

```diff
@@ -0,0 +1,405 @@
+  m_.FormSystemMatrix(empty, mass_bc_);
+
+  // Get the restriction and prolongation matrix for transformations
+  restriction_matrix_ = fespace->GetRestrictionMatrix();
+  prolongation_matrix_ = fespace->GetProlongationMatrix();
+
+  // Resize the the vectors B and X to the appropriate size
+  if (prolongation_matrix_) {
+    B_.SetSize(prolongation_matrix_->Width());
+  } else {
+    mfem::err << "<SPDESolver> prolongation matrix is not defined" << std::endl;
+  }
+  if (restriction_matrix_) {
+    X_.SetSize(restriction_matrix_->Height());
+  } else {
+    mfem::err << "<SPDESolver> restriction matrix is not defined" << std::endl;
+  }
+
+  sw.Stop();
+  if (Mpi::Root()) {
+    mfem::out << "<SPDESolver::Timing> matrix assembly " << sw.RealTime()
+              << " [s]" << std::endl;
+  }
+}
+
+void SPDESolver::Solve(ParLinearForm &b, ParGridFunction &x) {
+  // ------------------------------------------------------------------------
+  // Solve the PDE (A)^N g = f, i.e. compute g = (A)^{-1}^N f iteratively.
+  // ------------------------------------------------------------------------
+
+  StopWatch sw;
+  sw.Start();
+
+  // Zero initialize x to avoid touching uninitialized memory
+  x = 0.0;
+
+  ParGridFunction helper_gf(fespace_ptr_);
+  helper_gf = 0.0;
+
+  if (integer_order_of_exponent_ > 0) {
+    if (Mpi::Root()) {
+      mfem::out << "<SPDESolver> Solving PDE (A)^" << integer_order_of_exponent_
+                << " u = f" << std::endl;
+    }
+    ActivateRepeatedSolve();
+    Solve(b, helper_gf, 1.0, 1.0, integer_order_of_exponent_);
+    if (integer_order_) {
+      // If the exponent is an integer, we can directly add the solution to the
+      // final solution and return.
+      x += helper_gf;
+      if (!bc_.dirichlet_coefficients.empty()) {
+        LiftSolution(x);
+      }
+      return;
+    }
+    UpdateRHS(b);
+    DeactivateRepeatedSolve();
+  }
+
+  // ------------------------------------------------------------------------
+  // Solve the (remaining) fractional PDE by solving M integer order PDEs and
+  // adding up the solutions.
+  // ------------------------------------------------------------------------
+  if (!integer_order_) {
+    // Iterate over all expansion coefficient that contribute to the
+    // solution.
+    for (int i = 0; i < coeffs_.Size(); i++) {
+      if (Mpi::Root()) {
+        mfem::out << "\n<SPDESolver> Solving PDE -Δ u + " << -poles_[i]
+                  << " u = " << coeffs_[i] << " g " << std::endl;
+      }
+      helper_gf = 0.0;
+      Solve(b, helper_gf, 1.0 - poles_[i], coeffs_[i]);
+      x += helper_gf;
+    }
+  }
+
+  // Apply the inhomogeneous Dirichlet boundary conditions.
+  if (!bc_.dirichlet_coefficients.empty()) {
+    LiftSolution(x);
+  }
+
+  sw.Stop();
+  if (Mpi::Root()) {
+    mfem::out << "<SPDESolver::Timing> all PCG solves " << sw.RealTime()
+              << " [s]" << std::endl;
+  }
+}
+
+void SPDESolver::GenerateRandomField(ParGridFunction &x, int seed) {
```

Comment:
The behavior of this routine needs to be updated. (@psocratis has already implemented a fix.)

Currently, the seed gets reset every time an RF is generated. This means that repeated calls to the routine (with the same seed) will return the same field each time. There is no clear way to generate an identical *sequence* of RFs, which is what we would like.

A solution is to introduce a separate "SetSeed" function and no longer pass the seed to GenerateRandomField. The WhiteGaussianNoiseDomainLFIntegrator need only be created once in the SetSeed function and passed internally to the linear form b in GenerateRandomField.

Reply:
@psocratis If you have this fix ready, would you mind pushing it to this branch? :)